homeassistant/components/http/__init__.py

```diff
@@ -403,30 +431,58 @@ async def redirect(request: web.Request) -> web.StreamResponse:
             self.app.router.add_route("GET", url, redirect)
         )
 
-    def register_static_path(
-        self, url_path: str, path: str, cache_headers: bool = True
+    def _make_static_resources(
+        self, configs: Collection[StaticPathConfig]
+    ) -> dict[str, CachingStaticResource | web.StaticResource | None]:
+        """Create a list of static resources."""
+        return {
+            config.url_path: _STATIC_CLASSES[config.cache_headers](
+                config.url_path, config.path
+            )
+            if os.path.isdir(config.path)
+            else None
+            for config in configs
+        }
+
+    async def async_register_static_paths(
+        self, configs: Collection[StaticPathConfig]
     ) -> None:
         """Register a folder or file to serve as a static path."""
-        if os.path.isdir(path):
-            if cache_headers:
-                resource: CachingStaticResource | web.StaticResource = (
-                    CachingStaticResource(url_path, path)
-                )
-            else:
-                resource = web.StaticResource(url_path, path)
-            self.app.router.register_resource(resource)
-            self.app[KEY_ALLOW_CONFIGRED_CORS](resource)
-            return
+        resources = await self.hass.async_add_executor_job(
+            self._make_static_resources, configs
+        )
+        self._async_register_static_paths(configs, resources)
 
-        async def serve_file(request: web.Request) -> web.FileResponse:
-            """Serve file from disk."""
-            if cache_headers:
-                return web.FileResponse(path, headers=CACHE_HEADERS)
-            return web.FileResponse(path)
+    @callback
+    def _async_register_static_paths(
+        self,
+        configs: Collection[StaticPathConfig],
+        resources: dict[str, CachingStaticResource | web.StaticResource | None],
+    ) -> None:
+        """Register a folders or files to serve as a static path."""
+        app = self.app
+        allow_cors = app[KEY_ALLOW_CONFIGRED_CORS]
+        for config in configs:
+            if resource := resources[config.url_path]:
+                app.router.register_resource(resource)
+                allow_cors(resource)
+
+            target = (
+                _serve_file_with_cache_headers if config.cache_headers else _serve_file
+            )
+            allow_cors(
+                self.app.router.add_route(
+                    "GET", config.url_path, partial(target, config.path)
+                )
+            )
 
-        self.app[KEY_ALLOW_CONFIGRED_CORS](
-            self.app.router.add_route("GET", url_path, serve_file)
-        )
+    def register_static_path(
```

Comment:
This function needs to be deprecated because it does blocking I/O in the event loop.

A future PR will do that and add a report() here

Reply:
https://github.com/home-assistant/core/pull/119895